### ISSUE TITLE
Evaling from an .fnl file buffer

### ISSUE THREAD

**Issue by agzam** [warreq/friar#1]
Thank you for this package, it makes my life a bit less difficult. Even though I understand that this is an old project and you may not use it much anymore, if I may, I have two questions:

First, is it possible to eval things directly from a .fnl buffer? I'd love to select an expression and let awesomewm execute it.

Secondly, why doesn't it understand local vars? Why, when I type something like `(local gears (require :gears))` in the REPL buffer, doesn't it recognize that? I always have to make globals, i.e., `(global gears (require :gears))`. Perhaps I'm missing something? Maybe you wrote this project at the time when there was no [built-in repl in fennel](https://wiki.fennel-lang.org/Repl)? What would be the most sensible REPL setup today, making them all to work together - Fennel, Emacs and AwesomeWM?

**Comment by shtwzrd** [comment]
Howdy neighbor, happy to hear its been of use, thank you. :) Apologies for the slow response.

I recall @russmatney added [some features](https://github.com/russmatney/friar/commit/8b2e0a26a84f016d1b547095f67942a2ba743f02) to friar in a personal fork, including evaling expressions directly from .fnl buffers. I recommend using his fork over mine for that reason. (Side note @russmatney : are you still using Friar at all? I could update the README to point people to your repo in that case. If not, maybe I should just pull your changes so people are more likely to find your improvements. :) )

As for the local vars, that's quite an annoyance, I agree. The reason the locals do not get recognized is that we communicate with AwesomeWM's Lua VM via dbus, [which just calls `load()`](https://github.com/awesomeWM/awesome/blob/e6f5c7980862b7c3ec6c50c643b15ff2249310cc/lib/awful/remote.lua#L22).

Any globals declared in `code` there are going to pollute _ENV as a side effect, but locals only exist within the scope of the code passed to `load()`, so they no longer exist in your next invocation. You can observe the exact same behavior in `awesome-client`.

This is a common issue with Lua locals and REPLs -- Fennel's REPL implementation works around it [like this](https://git.sr.ht/~technomancy/fennel/tree/main/item/src/fennel/repl.fnl) if you're interested in finding a proper solution to it.

The workaround I had for this was to just declare globals for the things I frequently reference, like `gears`, `awful`, `beautiful`, etc, as well as some of my more common personal utility modules. Still annoying though.

I think if I were to remake Friar today, there's two routes I would consider:

- Have it communicate directly with a running Fennel REPL and re-use everything emacs-fennel-mode has to offer. Have some preamble that puts the awesome libs onto the loadpath, and forces it to load the code from awesome config, so it's a reasonable 'mirror' of the VM state of awesomewm. Then Friar's main job is just to write-back all the manipulations you do in the Fennel REPL back over to Awesome via D-BUS.

- Make some `friar-jack` function that injects fennel + [fennel-proto-repl](https://gitlab.com/andreyorst/fennel-proto-repl-protocol) implementation into awesomewm via d-bus, with a payload that sets up a proto-repl listening on a network port, and then connect fennel-proto-repl client to it. I would probably pursue this method first, as I think it'll be easier than trying to keep the LuaVM in sync.

**Comment by russmatney** [comment]
@shtwzrd thanks for the mention! I had intended to open a PR after battle-testing my changes a bit, but I didn't stick with them for long, so wasn't too confident in them yet. Feel free to pull the changes or prod me to open a PR, and I'll get to it soon!

I ended up working with awesomeWM from a clojure repl directly ([via awesome-client](https://github.com/russmatney/clawe/blob/3987390ffe538d878045e9d886190542fb111b9e/src/ralphie/awesome/fnl.clj#L96)), which led to building up an `awm-fnl` macro and a bunch of other hacks ([here's the awesomewm namespace](https://github.com/russmatney/clawe/blob/3987390ffe538d878045e9d886190542fb111b9e/src/ralphie/awesome.clj) and a [macro helper](https://github.com/russmatney/clawe/blob/3987390ffe538d878045e9d886190542fb111b9e/src/ralphie/awesome/fnl.clj) that declares globals, similarly to your described approach). It's not well documented and `clawe` is a big monorepo with quite a bit of scope at this point, so not too approachable... it's my pipe-dream to make it useful to more than just me at some point.

Anyway, just to share an alternate (tho somewhat over-engineered, haha) approach!
